x/stakeibc/keeper/icqcallbacks_validator_exchange_rate.go

```diff
@@ -37,26 +36,69 @@ func ValidatorExchangeRateCallback(k Keeper, ctx sdk.Context, args []byte, query
+	if query.HasTimedOut(ctx) {
+		if inLSMLiquidStakeCallback {
+			return k.LSMSlashQueryTimeout(ctx, hostZone, query)
+		}
+		return nil
+	}
+
 	// Unmarshal the query response args into a Validator struct
 	queriedValidator := stakingtypes.Validator{}
-	err := k.cdc.Unmarshal(args, &queriedValidator)
-	if err != nil {
+	if err := k.cdc.Unmarshal(args, &queriedValidator); err != nil {
 		return errorsmod.Wrapf(err, "unable to unmarshal query response into Validator type")
 	}
-	k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(chainId, ICQCallbackID_Validator, "Query response - Validator: %s, Jailed: %v, Tokens: %v, Shares: %v",
+	k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(chainId, ICQCallbackID_Validator,
+		"Query response - Validator: %s, Jailed: %v, Tokens: %v, Shares: %v",
 		queriedValidator.OperatorAddress, queriedValidator.Jailed, queriedValidator.Tokens, queriedValidator.DelegatorShares))
 
+	// Check the query response to identify if the validator was slashed
+	validatorWasSlashed, err := k.CheckIfValidatorWasSlashed(ctx, hostZone, queriedValidator)
+	if err != nil {
+		return err
+	}
+
+	// If we are in the LSMLiquidStake callback, finish the transaction
+	if inLSMLiquidStakeCallback {
+		if err := k.LSMSlashQueryCallback(ctx, hostZone, query, validatorWasSlashed); err != nil {
+			return errorsmod.Wrapf(err, "unable to finish LSM liquid stake")
+		}
+	}
+
+	// If the validator was slashed, we'll have to issue a delegator shares query to determine
+	// the magnitude of the slash
+	if validatorWasSlashed {
+		if err := k.QueryDelegationsIcq(ctx, hostZone, queriedValidator.OperatorAddress); err != nil {
+			return errorsmod.Wrapf(err, "Failed to submit ICQ validator delegations")
+		}
+	}
+
+	return nil
+}
+
+// Determines if the validator was slashed by comparing the validator exchange rate from the query response
+// with the exchange rate stored internally
+func (k Keeper) CheckIfValidatorWasSlashed(
+	ctx sdk.Context,
+	hostZone types.HostZone,
+	queriedValidator stakingtypes.Validator,
+) (validatorWasSlashed bool, err error) {
 	// Get the validator from the host zone
 	validator, valIndex, found := GetValidatorFromAddress(hostZone.Validators, queriedValidator.OperatorAddress)
 	if !found {
-		return errorsmod.Wrapf(types.ErrValidatorNotFound, "no registered validator for address (%s)", queriedValidator.OperatorAddress)
+		return false, errorsmod.Wrapf(types.ErrValidatorNotFound, "no registered validator for address (%s)", queriedValidator.OperatorAddress)
 	}
 	previousSharesToTokensRate := validator.InternalSharesToTokensRate
 
 	// If the validator's delegation shares is 0, we'll get a division by zero error when trying to get the exchange rate
 	//  because `validator.TokensFromShares` uses delegation shares in the denominator
 	if queriedValidator.DelegatorShares.IsZero() {
-		return errorsmod.Wrapf(types.ErrDivisionByZero,
+		return false, errorsmod.Wrapf(types.ErrDivisionByZero,
```

Comment:
This isn't a new issue with this PR but if I am understanding right, we won't be able to check if the validator was slashed or not if we don't have any shares there yet.  This is a problem because a 0 weight validator which we will still accept LSM tokens for could have 0 of our delegator shares on it; but then we won't be able to detect slashes and the users on that validator can then cash out the full amount to stTokens and be overpaid by the protocol... is this just currently unavoidable risk since we only detect slash via delegator shares changes?

Reply:
really good call out - I hadn't thought about that! 

To clarify thought, we detect slashes using the exchange rate query - irrespective of whether the number of delegated tokens changed - so we would be able to correctly identify that the validator was slashed. Additionally, zero-weight validators can have a delegated balance before a rebalancing.

So the concerning case is just the first delegation after a rebalancing to a zero-weight validator - in which case the RR is not adjusted

Reply:
You had asked if the LSM representing shares would fix this... I don't think it actually would because the core issue is that we don't have any shares on a given 0 weight val until someone comes in from out of the blue with an LSM liquid stake request.  In the shares representation, as long as we have an accurate exchange rate everything should be good... maybe for LSM coming from a 0 weight validator where we previously believed there to be zero shares we always trigger the second ICQ to get an updated exchange rate no matter how small the LSM stake amount is to account for this?

Reply:
So I think there's two main concerns here at the end of the day:
1. Are we minting the correct number of stTokens / was a user able to escape a slash?
2. Are we updating our redemption rate correctly?

For (1), if the exchange rate is updated, we always mint the correct number of stTokens now because the slash would have been reflected in the exchange rate.  If the exchange rate _is not_ updated, we do not mint the correct number of stTokens, but the magnitude of this error is capped by the checkpoint. 

For (2), the redemption rate is basically getting at the ratio of stTokens to native tokens. When calculating the number of native tokens, we need to sum the tokens that are currently staked. For zero-weight validators with respect to LSM, this can come from two sources: (a) the LSM liquid stakes that have not been detokenized yet, and (b) the liquid stakes that have been detokenized and is living as a native stake.

For (a), we can use the validator exchange rate to get the true value of that LSM token. For (b), we just want to make sure the amount we think we have delegated is correct. I believe if there was stake on that validator, the slash query should adjust it correctly. And if there was no stake on that validator, it's not important to the redemption rate equation.

Does that all track with you?

Reply:
@ethan-stride friendly bump on this, I think it's the last comment blocking the merge

Reply:
Yes, for (1) which is really what I was worried about, I think the key thing is that the magnitude of the error is capped by the checkpoint.  I was thinking about the zero weight validators being more likely to not have been updated recently but that shouldn't matter because the checkpoint sets a max cap on the potential error.  
For (2) I wasn't worried about the redemption rate so much as the exchange rate per validator.  If at some point we give out the "wrong" number of stTokens we still need to use the real number of stTokens in existence when doing all redemption rate computations going forward.  I don't disagree anywhere and we also had discussed part of this issue on the other related PRs, I'll close this issue.